src/textual/app.py

```diff
@@ -219,30 +234,36 @@ def log(
             _textual_calling_frame (inspect.FrameInfo | None): The frame info to include in
                 the log message sent to the devtools server.
         """
-        output = ""
-        try:
-            output = f" ".join(str(arg) for arg in objects)
-            if kwargs:
-                key_values = " ".join(f"{key}={value}" for key, value in kwargs.items())
-                output = " ".join((output, key_values))
-
-            if not _textual_calling_frame:
-                _textual_calling_frame = inspect.stack()[1]
-
-            calling_path = _textual_calling_frame.filename
-            calling_lineno = _textual_calling_frame.lineno
+        if verbosity > self.log_verbosity:
+            return
 
-            if self.devtools.is_connected and verbosity <= self.log_verbosity:
-                if len(objects) > 1 or len(kwargs) >= 1 and output:
-                    self.devtools.log(output, path=calling_path, lineno=calling_lineno)
-                else:
+        try:
+            if len(objects) == 1 and not kwargs:
+                if self._log_console is not None:
+                    self._log_console.print(objects[0])
+                if self.devtools.is_connected:
+                    if not _textual_calling_frame:
+                        _textual_calling_frame = inspect.stack()[1]
+                    calling_path = _textual_calling_frame.filename
+                    calling_lineno = _textual_calling_frame.lineno
                     self.devtools.log(
-                        *objects, path=calling_path, lineno=calling_lineno
+                        objects[0], path=calling_path, lineno=calling_lineno
                     )
-
-            if self.log_file and verbosity <= self.log_verbosity:
-                self.log_file.write(output + "\n")
-                self.log_file.flush()
+            else:
+                output = f" ".join(str(arg) for arg in objects)
```

Comment:
we likely don't need the f-string there?
(pylint detects this by default if I'm not wrong, it could be nice to add such a linter to the project at some point?)

Reply:
Good catch. I'm dubious about the benefits of such linters, I find the create a lot of busy work that doesn't improve the project. But if it were properly tuned, I could be persuaded!